In [1]:
import pandas as pd

#  Задание 1. Базовое изучение

In [2]:
df = pd.read_csv('C:\\Users\\kuznecov.ap\\Desktop\\piton\\15. Базовая статистика\\horse_data.csv',  sep = ',' , header=0, na_values='?')
df.columns = (['surgery?','Age', 'Hospital_Number','rectal_temperature','pulse', 'respiratory_rate', 'temperature_of_extremities', 'peripheral_pulse', 'mucous_membranes', 'capillary_refill_time', 'pain', 'peristalsis','abdominal_distension', 'nasogastric_tube','nasogastric_reflux', 'nasogastric_reflux PH','rectal_examination','abdomen', 'packed_cell_volume', 'total_protein','abdominocentesis_appearance', 'abdomcentesis_total_protein', 'outcome','surgical_lesion', 'type_of_lesion',  '25','26','cp_data'])                                                                      
# df.head(100)
new_df = df[['Age', 'rectal_temperature', 'pulse', 'respiratory_rate', 'pain', 'packed_cell_volume', 'outcome', 'surgical_lesion']]
for col in new_df.columns:
    pct_missing = new_df[col].isna().mean()
    print(f'{col} - {pct_missing :.1%}')
new_df.describe()

Age - 0.0%
rectal_temperature - 20.1%
pulse - 8.0%
respiratory_rate - 19.4%
pain - 18.4%
packed_cell_volume - 9.7%
outcome - 0.3%
surgical_lesion - 0.0%


Age  rectal_temperature       pulse  respiratory_rate  \
count  299.00000          239.000000  275.000000        241.000000   
mean     1.64214           38.166527   71.934545         30.427386   
std      2.17730            0.733508   28.680522         17.678256   
min      1.00000           35.400000   30.000000          8.000000   
25%      1.00000           37.800000   48.000000         18.000000   
50%      1.00000           38.200000   64.000000         24.000000   
75%      1.00000           38.500000   88.000000         36.000000   
max      9.00000           40.800000  184.000000         96.000000   

             pain  packed_cell_volume     outcome  surgical_lesion  
count  244.000000          270.000000  298.000000       299.000000  
mean     2.942623           46.300000    1.550336         1.361204  
std      1.303993           10.438384    0.737967         0.481155  
min      1.000000           23.000000    1.000000         1.000000  
25%      2.000000           38.000000    1.000000         1.000000  
50%      3.000000           45.000000    1.000000         1.000000  
75%      4.000000           52.000000    2.000000         2.000000  
max      5.000000           75.000000    3.000000         2.000000

1. загружен файл
2. замена '?' на 'NaN'
3. Подсчитаны пропуски и выведены базовые метрики
4. Выбраны 8 столбцов:
4.1 Age (Возраст (1-взрослая/2-молодая)) категориальная
4.2 rectal_temperature (Ректальная температура (норм 37,8)) числовая
4.3 pulse (Пульс) числовая
4.4 respiratory_rate (Частота дыхания (норм 8-10)) числовая
4.5 pain (боль(1 -норм/5 -непрерывная сильная боль)) категориальная
4.6 packed_cell_volume (объем упакованных клеток(30-50-норм)) числовая
4.7 outcome (гибель (1-жив, 2-мертв, 3-усыплен))) категориальная
4.8 surgical_lesion? (хирургическое повреждение?(1 -да/2 -нет)) категориальная

# Задание 2. Работа с выбросами 

In [3]:
q1 = new_df.surgical_lesion.quantile(0.25)
q3 = new_df.surgical_lesion.quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) 
upper_bound = q3 + (1.5 * iqr)
remove_outliers = new_df[new_df.surgical_lesion.between(lower_bound, upper_bound, inclusive=True)].sort_values('pulse')
remove_outliers
# print(lower_bound, upper_bound)

Age  rectal_temperature  pulse  respiratory_rate  pain  \
129    1                37.8   30.0              12.0   NaN   
231    1                38.5   30.0              18.0   NaN   
275    1                38.0   36.0              16.0   4.0   
202    1                37.2   36.0               9.0   2.0   
241    1                 NaN   36.0              32.0   5.0   
..   ...                 ...    ...               ...   ...   
203    1                39.2    NaN              23.0   4.0   
206    1                 NaN    NaN               NaN   NaN   
215    1                 NaN    NaN               NaN   NaN   
226    9                 NaN    NaN               NaN   NaN   
287    1                 NaN    NaN               NaN   NaN   

     packed_cell_volume  outcome  surgical_lesion  
129                 NaN      2.0                1  
231                40.0      1.0                1  
275                37.0      3.0                2  
202                35.0      1.0                2  
241                41.0      2.0                1  
..                  ...      ...              ...  
203                36.0      1.0                1  
206                45.0      2.0                1  
215                 NaN      1.0                2  
226                45.0      1.0                1  
287                 NaN      1.0                1  

[299 rows x 8 columns]

In [4]:
print(remove_outliers.Age.mean())
print(remove_outliers.Age.median())
print(new_df.Age.round().mode()[0])

1.6421404682274248
1.0
1


- Age (Возраст): всего значений 299 из них выбросов 24. Выбросы удалить, т.к. значения по умолчанию 1 или 2. 
- rectal_temperature (Ректальная температура): всего значений 239 из них выбросов 14. Выбросы можно проигнорировать, т.к. разброс незначительный
- pulse (Пульс): всего значений 275 из них выбросов 5. Выбросы удалить, т.к. std=28.680522 и отклониния выбрсов значительные
- respiratory_rate (Частота дыхания): всего значений 241 из них выбросов 17. Выбросы удалить, т.к. std=17.678256 и отклониния выбрсов значительные
- pain (боль): всего значений 244 из них выбросов 0. 
- packed_cell_volume (объем упакованных клеток): всего значений 270 из них выбросов 3. Выбросы удалить, т.к. std=46.300000 и отклониния выбрсов значительные
- outcome (гибель): всего значений 298 из них выбросов 0
- surgical_lesion? (хирургическое повреждение?): всего значений 299 из них выбросов 0

# Задание 3. Работа с пропусками


In [4]:
new_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 299 entries, 0 to 298
Data columns (total 8 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Age                 299 non-null    int64  
 1   rectal_temperature  239 non-null    float64
 2   pulse               275 non-null    float64
 3   respiratory_rate    241 non-null    float64
 4   pain                244 non-null    float64
 5   packed_cell_volume  270 non-null    float64
 6   outcome             298 non-null    float64
 7   surgical_lesion     299 non-null    int64  
dtypes: float64(6), int64(2)
memory usage: 18.8 KB


In [14]:
for col in new_df.columns:
    pct_missing = new_df[col].isna().mean()
    print(f'{col} - {pct_missing :.1%}')


Age - 0.0%
rectal_temperature - 20.1%
pulse - 8.0%
respiratory_rate - 19.4%
pain - 0.0%
packed_cell_volume - 9.7%
outcome - 0.0%
surgical_lesion - 0.0%


Предполагаем, что пропуски носят случайный характер, поэтому предлагаю для числовых переменных произвести замену на медиану а для категориальных на моду (хотя в лекции Вы не рекомендовали так делать, но на данном этапе трудно выявить взаимосвязь переменных между собой).
- Age (Возраст): Пропусков нет 
- rectal_temperature (Ректальная температура): числовая переменная и 20,1% пропусков- заменяем на медиану
- pulse (Пульс): числовая переменная и 8% пропусков- заменяем на медиану
- respiratory_rate (Частота дыхания): числовая переменная и 19,4% пропусков- заменяем на медиану
- pain (боль):  категориальная переменная и 18,4% пропусков- заменяем на моду
- packed_cell_volume (объем упакованных клеток): числовая переменная и 9,7% пропусков- заменяем на медиану
- outcome (гибель): категориальная переменная и 0,3% пропусков- заменяем на моду
- surgical_lesion? (хирургическое повреждение?): Пропусков нет

In [26]:
new_df.rectal_temperature.fillna(new_df.rectal_temperature.median(), inplace=True)
new_df.rectal_temperature.fillna(new_df.rectal_temperature.median()).isna().sum()
new_df.pulse.fillna(new_df.pulse.median(), inplace=True)
new_df.pulse.fillna(new_df.pulse.median()).isna().sum()
new_df.respiratory_rate.fillna(new_df.respiratory_rate.median(), inplace=True)
new_df.respiratory_rate.fillna(new_df.respiratory_rate.median()).isna().sum()
new_df.pain.fillna(new_df.pain.mode()[0], inplace=True)
new_df.packed_cell_volume.fillna(new_df.packed_cell_volume.median(), inplace=True)
new_df.packed_cell_volume.fillna(new_df.packed_cell_volume.median()).isna().sum()
new_df.outcome.fillna(new_df.outcome.mode()[0], inplace=True)

In [27]:
new_df.rectal_temperature.fillna(new_df.rectal_temperature.median()).isna().sum()

0

In [28]:
print(new_df.rectal_temperature.mean())
print(new_df.rectal_temperature.median())
print(new_df.rectal_temperature.mode()[0])

print(new_df.rectal_temperature.std())
print(new_df.rectal_temperature.var())

38.17324414715723
38.2
38.2
0.6556567310086665
0.4298857489169709


In [29]:
for col in new_df.columns:
    pct_missing = new_df[col].isna().mean()
    print(f'{col} - {pct_missing :.1%}')

Age - 0.0%
rectal_temperature - 0.0%
pulse - 0.0%
respiratory_rate - 0.0%
pain - 0.0%
packed_cell_volume - 0.0%
outcome - 0.0%
surgical_lesion - 0.0%


In [30]:
new_df.describe()

Age  rectal_temperature       pulse  respiratory_rate  \
count  299.00000          299.000000  299.000000        299.000000   
mean     1.64214           38.173244   71.297659         29.180602   
std      2.17730            0.655657   27.586016         16.067835   
min      1.00000           35.400000   30.000000          8.000000   
25%      1.00000           37.900000   48.000000         20.000000   
50%      1.00000           38.200000   64.000000         24.000000   
75%      1.00000           38.500000   88.000000         34.500000   
max      9.00000           40.800000  184.000000         96.000000   

             pain  packed_cell_volume     outcome  surgical_lesion  
count  299.000000          299.000000  299.000000       299.000000  
mean     2.953177           46.173913    1.548495         1.361204  
std      1.177736            9.924963    0.737415         0.481155  
min      1.000000           23.000000    1.000000         1.000000  
25%      2.000000           39.000000    1.000000         1.000000  
50%      3.000000           45.000000    1.000000         1.000000  
75%      4.000000           50.000000    2.000000         2.000000  
max      5.000000           75.000000    3.000000         2.000000